### PR TITLE
Make composition associated with QgsProject

### PR DESCRIPTION
**Pull request by wonder-sk**
... rather than using project singleton within compositions

### CODE REVIEW

File: src/core/composer/qgsaddremoveitemcommand.cpp [hunk 1]
```diff
@@ -80,5 +80,5 @@ void QgsAddRemoveItemCommand::switchState()
     emit itemAdded( mItem );
     mState = Added;
   }
-  QgsProject::instance()->setDirty( true );
+  mComposition->project()->setDirty( true );
```
Comment (nyalldawson) [inline src/core/composer/qgsaddremoveitemcommand.cpp]: I'm wondering - should we guard all these project() uses against null pointers and allow compositions to be created without an associated project? I'm thinking here things like creating a QgsComposition from a template without it being linked to a specific project.

**Comment by nyalldawson** [comment]
Another thought - in order for compositions to be usable when multiple projects can exist we're going to need some type of QgsCompositionManager class attached to QgsProject which stores and handles the compositions attached to that project.... At the moment compositions are attached to the application (via composer windows!).

File: src/core/composer/qgsaddremoveitemcommand.cpp (hunk 1 above)
Comment (wonder-sk) [inline src/core/composer/qgsaddremoveitemcommand.cpp]: Actually I wanted to make QgsProject association optional, but in the end scratched the idea for now. The reason is that I would need to add a bunch of extra logic to handle cases with no project and more logic when project association is changed. Also, working without project is fine various cases, but in many other cases the composition would misbehave (e.g. access to project's map layers in map item, scale bar's distance calculator config from project, legend uses project's layer tree by default). So for now I decided to always force project association with composer, and later we may decide to relax that and make project association optional... does that make sense?

**Comment by wonder-sk** [comment]
Yeah the fact that compositions are attached to project via composer windows in QGIS app code is quite frustrating. I am not sure whether I will have time and energy to change that within the project refactoring project - there is still great amount of work pending just to make core library independent from QgsProject singleton..

File: src/core/composer/qgsaddremoveitemcommand.cpp (hunk 1 above)
Comment (nyalldawson) [inline src/core/composer/qgsaddremoveitemcommand.cpp]: Yep... There's nothing in this API forcing the project use, so happy to say project is mandatory for now and address this when required!

**Comment by nyalldawson** [comment]
> I am not sure whether I will have time and energy to change that within the project refactoring project - there is still great amount of work pending just to make core library independent from QgsProject singleton..

Fair enough! I actually misread this PR description at first and thought this was what the PR did, so got all excited. These changes are very welcome in any case! 

Composition manager has been on my to-do list for ages. Maybe the server team will tackle this when they do away with the server project parsers and need access to compositions for the server getprint functionality...?

**Comment by wonder-sk** [comment]
Yeah I hope server team could look into that eventually - though even for server team it is probably not the highest priority as one can still load compositions separately by listening to readProject() signal...

**Pull request merged**